src/storage/src/hummock/sstable/forward_sstable_iterator.rs

```diff
@@ -32,6 +37,109 @@ pub trait SstableIteratorType: HummockIterator + 'static {
         read_options: Arc<SstableIteratorReadOptions>,
     ) -> Self;
 }
+enum BlockFetcher {
+    Simple,
+    Prefetch(PrefetchContext),
+}
+
+impl BlockFetcher {
+    async fn get_block(
+        &mut self,
+        sst: &Sstable,
+        block_idx: usize,
+        sstable_store: &SstableStore,
+        stats: &mut StoreLocalStatistic,
+    ) -> HummockResult<BlockHolder> {
+        match self {
+            BlockFetcher::Simple => {
+                sstable_store
+                    .get(sst, block_idx, crate::hummock::CachePolicy::Fill, stats)
+                    .await
+            }
+            BlockFetcher::Prefetch(context) => {
+                context
+                    .get_block(sst, block_idx, sstable_store, stats)
+                    .await
+            }
+        }
+    }
+}
+
+struct PrefetchContext {
+    prefetched_blocks: VecDeque<(usize, BlockResponse)>,
+
+    /// block[cur_idx..=dest_idx] will definitely be visited in the future.
+    dest_idx: usize,
+}
+
+impl PrefetchContext {
+    fn new(dest_idx: usize) -> Self {
+        Self {
+            prefetched_blocks: VecDeque::with_capacity(2),
+            dest_idx,
+        }
+    }
+
+    async fn get_block(
+        &mut self,
+        sst: &Sstable,
+        idx: usize,
+        sstable_store: &SstableStore,
+        stats: &mut StoreLocalStatistic,
+    ) -> HummockResult<BlockHolder> {
+        let in_prefetch = {
```

Comment:
> To make the codes clearer, I suggest we decouple the logic to get `idx` block and prefetch like this:
> 
> ```
> // Get block with `idx`
> let block_response = if let Some((prefetched_idx, _)) = self.prefetched_blocks.front() {
>     if *prefetched_idx == idx {
>         self.prefetched_blocks.pop_front().unwrap().1
>     } else {
>         tracing::warn!(...);
>         self.prefetched_blocks.clear();
>         sstable_store
>             .get_block_response(sst, idx, crate::hummock::CachePolicy::Fill, stats).await?
>     }
> }
> 
> // Prefetch
> let next_prefetch_idx = ...
> if next_prefetch_idx <= self.dest_idx {
>     self.prefetched_blocks.push_back((
>         next_prefetch_idx,
>         sstable_store
>             .get_block_response(
>                 sst,
>                 next_prefetch_idx,
>                 crate::hummock::CachePolicy::Fill,
>                 stats,
>             )
>             .await?,
>     ));
> }
> 
> // Return
> block_response.wait().await
> ```

As I said before, A slight difference is that when prefetch policy is on while no blocks need prefetching, this PR manages to call the original `get` interface but the modified version here will call `prefetch` interface. Though these two interfaces almost do the same thing at present, it will prevent further optimization.

Reply:
> To make the codes clearer, I suggest we decouple the logic to get `idx` block and prefetch like this:
> 
> ```
> // Get block with `idx`
> let block_response = if let Some((prefetched_idx, _)) = self.prefetched_blocks.front() {
>     if *prefetched_idx == idx {
>         self.prefetched_blocks.pop_front().unwrap().1
>     } else {
>         tracing::warn!(...);
>         self.prefetched_blocks.clear();
>         sstable_store
>             .get_block_response(sst, idx, crate::hummock::CachePolicy::Fill, stats).await?
>     }
> }
> 
> // Prefetch
> let next_prefetch_idx = ...
> if next_prefetch_idx <= self.dest_idx {
>     self.prefetched_blocks.push_back((
>         next_prefetch_idx,
>         sstable_store
>             .get_block_response(
>                 sst,
>                 next_prefetch_idx,
>                 crate::hummock::CachePolicy::Fill,
>                 stats,
>             )
>             .await?,
>     ));
> }
> 
> // Return
> block_response.wait().await
> ```

As I said before, a slight difference is that when prefetch policy is on but this is the last block (so no need to prefetch), this PR manages to call the original `get` interface but the modified version here will call `prefetch` interface. Though these two interfaces almost do the same thing at present, it will prevent further optimization.

Reply:
Do you mean that `sstable_store.get` and `sstable_store.get_block_response` will have different implementations in the future? What further optimizations are you thinking of?

Regardless, we can still call the original `get` interface by checking whether the provided `idx` is the last block with get and prefetch decoupled. It is better to organize the codes in a cleaner and equivalent way instead of mixing the logics together with branchings.

Reply:
> Do you mean that `sstable_store.get` and `sstable_store.get_block_response` will have different implementations in the future? What further optimizations are you thinking of?
> 
> Regardless, we can still call the original `get` interface by checking whether the provided `idx` is the last block with get and prefetch decoupled. It is better to organize the codes in a cleaner and equivalent way instead of mixing the logics together with branchings.

ok, fixed